raft/snapshot.go

```diff
@@ -1,66 +1,253 @@
+		raftId := uint16(rawRaftId)
+
+		if raftId == pm.raftId {
+			snapshot.addresses[i] = *pm.address
+		} else {
+			snapshot.addresses[i] = *pm.peers[raftId].address
+		}
 	}
+	sort.Sort(ByRaftId(snapshot.addresses))
 
-	return pm.wal.ReleaseLockTo(snap.Metadata.Index)
-}
+	// Populate removed IDs
 
-func (pm *ProtocolManager) maybeTriggerSnapshot() {
-	if pm.appliedIndex-pm.snapshotIndex < snapshotPeriod {
-		return
+	for i, removedIface := range pm.removedPeers.List() {
+		snapshot.removedRaftIds[i] = removedIface.(uint16)
 	}
 
-	pm.triggerSnapshot()
+	return snapshot
 }
 
-func (pm *ProtocolManager) triggerSnapshot() {
-	glog.V(logger.Info).Infof("start snapshot [applied index: %d | last snapshot index: %d]", pm.appliedIndex, pm.snapshotIndex)
-	snapData := pm.blockchain.CurrentBlock().Hash().Bytes()
-	snap, err := pm.raftStorage.CreateSnapshot(pm.appliedIndex, &pm.confState, snapData)
+// Note that we do *not* read `pm.appliedIndex` here. We only use the `index`
+// parameter instead. This is because we need to support a scenario when we
+// snapshot for a future index that we have not yet recorded in LevelDB. See
+// comments around the use of `forceSnapshot`.
+func (pm *ProtocolManager) triggerSnapshot(index uint64) {
+	pm.mu.RLock()
+	snapshotIndex := pm.snapshotIndex
+	pm.mu.RUnlock()
+
+	glog.V(logger.Info).Infof("start snapshot [applied index: %d | last snapshot index: %d]", index, snapshotIndex)
+
+	snapData := pm.buildSnapshot().toBytes()
+	snap, err := pm.raftStorage.CreateSnapshot(index, &pm.confState, snapData)
 	if err != nil {
 		panic(err)
 	}
-	if err := pm.saveSnapshot(snap); err != nil {
+	if err := pm.saveRaftSnapshot(snap); err != nil {
 		panic(err)
 	}
-	// Discard all log entries prior to appliedIndex.
-	if err := pm.raftStorage.Compact(pm.appliedIndex); err != nil {
+	// Discard all log entries prior to index.
+	if err := pm.raftStorage.Compact(index); err != nil {
 		panic(err)
 	}
-	glog.V(logger.Info).Infof("compacted log at index %d", pm.appliedIndex)
-	pm.snapshotIndex = pm.appliedIndex
+
+	glog.V(logger.Info).Infof("compacted log at index %d", index)
+
+	pm.mu.Lock()
+	pm.snapshotIndex = index
+	pm.mu.Unlock()
+}
+
+func confStateIdSet(confState raftpb.ConfState) *set.Set {
+	set := set.New()
+	for _, rawRaftId := range confState.Nodes {
+		set.Add(uint16(rawRaftId))
+	}
+	return set
 }
 
-// For persisting cluster membership changes correctly, we need to trigger a
-// snapshot before advancing our persisted appliedIndex in LevelDB.
-//
-// See handling of EntryConfChange entries in raft/handler.go for details.
-func (pm *ProtocolManager) triggerSnapshotWithNextIndex(index uint64) {
-	pm.appliedIndex = index
-	pm.triggerSnapshot()
+func (pm *ProtocolManager) updateClusterMembership(newConfState raftpb.ConfState, addresses []Address, removedRaftIds []uint16) {
+	glog.V(logger.Info).Infof("updating cluster membership per raft snapshot")
+
+	prevConfState := pm.confState
+
+	// Update tombstones
```

Comment:
This would also be worth a note. The long discussion we had about how to remove peers / why we should tombstone has already mostly faded from memory.